src/mnode/mnode-pastel.cpp

```diff
@@ -1436,6 +1435,168 @@ std::unique_ptr<CPastelTicket> CArtTradeTicket::FindArtRegTicket() const
     return std::move(chain.front());
 }
 
+// CArtRoyaltyTicket ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+CArtRoyaltyTicket CArtRoyaltyTicket::Create(
+    std::string _pastelID, std::string _newPastelID,
+    std::string _artTnxId, const SecureString& strKeyPass) {
+  CArtRoyaltyTicket ticket(std::move(_pastelID), std::move(_newPastelID));
+
+  ticket.artTnxId = std::move(_artTnxId);
+
+  ticket.GenerateTimestamp();
+
+  std::string strTicket = ticket.ToStr();
+  ticket.signature = CPastelID::Sign(reinterpret_cast<const unsigned char*>(strTicket.c_str()), strTicket.size(), ticket.pastelID, strKeyPass);
+
+  return ticket;
+}
+
+std::string CArtRoyaltyTicket::ToStr() const noexcept {
+  std::stringstream ss;
+  ss << pastelID;
+  ss << newPastelID;
+  ss << artTnxId;
+  ss << m_nTimestamp;
+  return ss.str();
+}
+
+bool CArtRoyaltyTicket::IsValid(std::string& errRet, bool preReg, int depth) const {
+  unsigned int chainHeight = 0; {
+    LOCK(cs_main);
+    chainHeight = static_cast<unsigned int>(chainActive.Height()) + 1;
+  }
+
+  // 0. Common validations
+  std::unique_ptr<CPastelTicket> pastelTicket;
+  if (!common_validation(*this, preReg, artTnxId, pastelTicket,
+      [](const TicketID tid) { return (tid != TicketID::Art); },
+      "Royalty", "art", depth, TicketPrice(chainHeight), errRet)) {
+    return false;
+  }
+
+  // Check the Royalty change ticket for that Art is already in the database
+  // (ticket transaction replay attack protection)
+  CArtRoyaltyTicket _ticket;
+  if (FindTicketInDb(KeyOne(), _ticket) &&
+     (preReg ||  // if pre reg - this is probably repeating call, so signatures can be the same
+      _ticket.signature != signature || !_ticket.IsBlock(m_nBlock) || _ticket.m_txid != m_txid)) {
+    errRet = strprintf("Change Royalty ticket is already registered in blockchain [pastelID = %s; new_pastelID = %s]"
+                       "[this ticket block = %d txid = %s; found ticket block = %d txid = %s] with art txid [%s]",
+                       pastelID, newPastelID, m_nBlock, m_txid, _ticket.GetBlock(), _ticket.m_txid, artTnxId);
+    return false;
+  }
+
+  if (newPastelID.empty()) {
+    errRet = strprintf("Change Royalty ticket new_pastelID is empty");
```

Comment:
you don't need to use strprintf here - nothing to format, just string assignment is needed here

Reply:
ok